### PR TITLE
Reduce linked node size in Enumerable.Append/Prepend/Union

### PR DESCRIPTION
**Pull request by jamesqo**
- For `Append` / `Prepend` the count of previous iterators isn't needed by this iterator, so we can move that stuff from the nodes to the iterators which can be GC'd on successive linked Appends/Prepends.

- Similarly, all of the non-readonly state for `Union` of previous iterators isn't relevant for this iterator, so use `SingleLinkedNode` to only keep alive what we need (the sources). Additionally, all of the iterators have the same comparer so we only need to store the comparer in the latest node.

- Introduce `Set.UnionWith` for brevity and move `SingleLinkedNode` into its own file.

**edit:** Was finally able to run [perf tests](https://gist.github.com/jamesqo/2039702fe841a5264dd73b47c6e9efd5) for this. GCs have increased, as expected, because there is a substantial reduction in leaked memory. `Concat` and `Union` see 50-60% reductions in leaked memory, `Append` and `Prepend` about 20%.

**edit 2:** Ignore the results for `Concat`, I hit a few complications there and the changes became pretty big. I'm going to save them for another PR.

/cc @JonHanna @VSadov

### CODE REVIEW

**Comment by jamesqo** [comment]
Note: I tried to do this for `Concat` too, where we also have a linked-list scheme similar to that of `Union`. However, it isn't possible at the moment because [this part](https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Concat.cs#L321) where we first append an `IEnumerable` to a list of `ICollection`s would require converting a `SingleLinkedNode<ICollection<TSource>>` to a `SingleLinkedNode<IEnumerable<TSource>>`, and classes are not variant. The only way in which we could do that would be to do an O(N) close of the whole list, which seems prohibitively expensive for long chains.

I suspect, however, that when https://github.com/dotnet/corefx/pull/14675 goes in then `Concat2/NCollectionIterator` will have lost much of their merit because using `SparseArrayBuilder` we will still be able to copy the collections to an array without buffering/resizing. Then we will be able to remove those types, treat everything as an `IEnumerable<TSource>`, and use `SingleLinkedNode` there.

File: src/System.Linq/src/System/Linq/Concat.cs [hunk 1]
```diff
@@ -81,327 +73,64 @@ internal override IEnumerable<TSource> GetEnumerable(int index)
...
-                ConcatNEnumerableIterator<TSource> current = this;
-                while (true)
-                {
-                    if (index == current._nextIndex)
-                    {
-                        return current._next;
-                    }
-
-                    ConcatIterator<TSource> previous = current._previousConcat;
+            internal override bool AllSourcesAreCollections => _headIndex < 0;
```
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Concat.cs]: Self-note: I think this can be moved into its own field later, since it shouldn't increase field size.

**Comment by jamesqo** [comment]
OK, I was finally able to post perf tests! :tada:

@JonHanna, @VSadov PTAL

@karelz This can be un-marked as no merge.

**Comment by JonHanna** [comment]
I'm honestly not sure whether to consider this an improvement or not.

I'd be inclined to favour what benefits `Concat` and `Union` over what favours `Append` and `Prepend` and this seems to do the opposite in terms of speed if I'm reading the performance tests correctly. But maybe I shouldn't think that way (how much can we perhaps expect heavy use of `Append` and `Prepend` in the future?).

They all benefit in terms of memory, but if the objects produced were short-lived (which is v. common with linq) it would all be collected anyway, no? I'm not sure how to judge the value of something that saves in that regard but increases GC calls.

I does seem to have simplified things a bit, and that's always a good thing.

**Comment by jamesqo** [comment]
> I'd be inclined to favour what benefits Concat and Union over what favours Append and Prepend and this seems to do the opposite in terms of speed if I'm reading the performance tests correctly.

The performance tests were only written to illustrate the impact on memory retention, not CPU time. It does not actually iterate any of the enumerables returned from the Linq methods. Iteration would quickly overshadow any impact on the time it takes to create the iterators.

I have encountered significant speed/memory regressions when calling Count/ToArray/ToList on Concat iterators, though, so you are correct in that sense. This is mostly because there are extra virtual calls and unnecessary re-walks of the linked list via `GetEnumerable`. I'm deciding to save those changes for a later PR because it's a lot more code and I'm no longer sure deleting the iterators is a clear win.

> They all benefit in terms of memory, but if the objects produced were short-lived (which is v. common with linq) it would all be collected anyway, no? I'm not sure how to judge the value of something that saves in that regard but increases GC calls.

I actually thought that allowing older iterators to be GC'd was the reason you used `SingleLinkedNode` in your original Append/Prepend implementation, instead of putting the items directly in the iterator. Was that the case or did you do it for some other reason?

BTW, there is also a benefit (albeit small) in terms of CPU time here too: since nodes have uniform type we no longer have to typecast when walking the linked list, since the tail is just represented as a null. https://github.com/dotnet/corefx/pull/15389/files#diff-49756bcd3a3a836bf3963f982c6af0f8L253

**Comment by JonHanna** [comment]
Ah. I see where you're going now.

So all that's left from what I've said above is that it seems to have simplified things, which is good 😄

**Comment by JonHanna** [comment]
Looking at the results at first I thought perhaps it was forcing more GC rather than allowing sooner GC. Seeing it's the latter, this seems like good stuff. I like the unionwith added, too. Nothing seems awry, so it's all down to the perf results really, do LGTM.

**Comment by karelz** [comment]
> @jamesqo  This can be un-marked as no merge.

Done

**Pull request merged**